src/main/policy.c

```diff
@@ -52,26 +52,42 @@
 
 #define POLICY_UPDATE() *policy_p = policy;
 
+// TODO: Python exceptions should be propagated up instead of being cleared
+// but the policy helper functions don't handle this case and they only populate
+// an as_error object and return a status code.
+// That will take too much time to refactor, so just clear the exception and
+// populate the as_error object instead. This currently makes it harder to
+// debug why a C-API call failed though, because we don't have the exact
+// exception that was thrown
 #define POLICY_SET_FIELD(__field, __type)                                      \
     {                                                                          \
-        PyObject *py_field = PyDict_GetItemString(py_policy, #__field);        \
-        if (py_field) {                                                        \
-            if (PyLong_Check(py_field)) {                                      \
-                policy->__field = (__type)PyLong_AsLong(py_field);             \
-            }                                                                  \
-            else {                                                             \
-                return as_error_update(err, AEROSPIKE_ERR_PARAM,               \
-                                       "%s is invalid", #__field);             \
-            }                                                                  \
+        PyObject *py_field_name = PyUnicode_FromString(#__field);              \
+        if (py_field_name == NULL) {                                           \
+            PyErr_Clear();                                                     \
+            return as_error_update(err, AEROSPIKE_ERR_CLIENT,                  \
+                                   "Unable to create Python unicode object");  \
         }                                                                      \
-    }
-
-#define POLICY_SET_BASE_FIELD(__field, __type)                                 \
-    {                                                                          \
-        PyObject *py_field = PyDict_GetItemString(py_policy, #__field);        \
+        PyObject *py_field =                                                   \
+            PyDict_GetItemWithError(py_policy, py_field_name);                 \
+        if (py_field == NULL && PyErr_Occurred()) {                            \
+            PyErr_Clear();                                                     \
+            Py_DECREF(py_field_name);                                          \
+            return as_error_update(                                            \
+                err, AEROSPIKE_ERR_CLIENT,                                     \
+                "Unable to fetch field from policy dictionary");               \
+        }                                                                      \
+        Py_DECREF(py_field_name);                                              \
```

Comment:
Why not put `Py_DECREF(py_field_name);` before the if?

Reply:
I didn't want to call Py_DECREF after an error indicator was set:

https://docs.python.org/3/c-api/exceptions.html

> If the error is not handled or carefully propagated, additional calls into the Python/C API may not behave as intended and may fail in mysterious ways.

Reply:
that makes sense, I was not aware of this 😅